File: java/DesignPatterns/CreationalDesignPatterns/SingletonDesignPattern/SingletonDesignPattern.Example5/src/main/java/creational/singleton/example5/Singleton.java
package creational.singleton.example5;

// Initialization-on-demand holder idiom — Java's equivalent of Lazy<T>.
// The inner static class is only loaded when getInstance() is first called,
// providing lazy initialization that is also thread-safe without synchronization.
public class Singleton {

    private Singleton() { }

    private static class SingletonHolder {
        private static final Singleton INSTANCE = new Singleton();
    }

    public static Singleton getInstance() {
        return SingletonHolder.INSTANCE;
    }
}

File: nodejs/DesignPatterns/CreationalDesignPatterns/SingletonDesignPattern/SingletonDesignPattern.Example5/src/Singleton.ts
import { Lazy } from './Lazy';

// Equivalent of C#: private static readonly Lazy<Singleton> _lazyInstance = new(() => new Singleton());
export class Singleton {
  private static readonly _lazyInstance = new Lazy<Singleton>(() => new Singleton());

  private constructor() {}

  static get instance(): Singleton {
    return Singleton._lazyInstance.value;
  }
}
